src/main/java/pl/llp/aircasting/screens/common/helpers/LocationHelper.java

```diff
@@ -20,153 +20,150 @@
-        updateLocation(location);
+    @SuppressLint("MissingPermission")
+    public void initLocation() {
+        mFusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(mContext);
+
+        mFusedLocationProviderClient.getLastLocation()
+                .addOnSuccessListener(new OnSuccessListener<Location>() {
+                    @Override
+                    public void onSuccess(Location location) {
+                        if (location != null) {
+                            mLastLocation = location;
+                        }
+                    }
+                });
     }
 
-    private void updateLocation(Location location) {
-        if (isBetterLocation(location, lastLocation)) {
-            lastLocation = location;
+    @SuppressLint("MissingPermission")
+    public synchronized void startLocationUpdates() {
+        mFusedLocationProviderClient.requestLocationUpdates(mLocationRequest, mLocationCallback, null);
+        mLocationUpdatesStarted = true;
+    }
 
-            LocationEvent locationEvent = new LocationEvent();
-            eventBus.post(locationEvent);
+    public void stopLocationUpdates() {
+        if (mFusedLocationProviderClient != null && locationUpdatesStarted()) {
+            mFusedLocationProviderClient.removeLocationUpdates(mLocationCallback);
+            mLocationUpdatesStarted = false;
         }
     }
 
-    @Override
-    public void onStatusChanged(String s, int i, Bundle bundle) {
+    private boolean checkLocationSettingsSatisfied(final Activity activity) {
+        mLocationRequest = createLocationRequest();
+        LocationSettingsRequest.Builder builder = new LocationSettingsRequest.Builder()
+                .addLocationRequest(mLocationRequest);
+        SettingsClient client = LocationServices.getSettingsClient(mContext);
+        Task<LocationSettingsResponse> task = client.checkLocationSettings(builder.build());
+
+        task.addOnSuccessListener(new OnSuccessListener<LocationSettingsResponse>() {
+            @Override
+            public void onSuccess(LocationSettingsResponse locationSettingsResponse) {
+                startLocationUpdates();
+                mLocationRequestListener.onLocationRequestSuccess();
+            }
+        });
+
+        task.addOnFailureListener(new OnFailureListener() {
+            @Override
+            public void onFailure(@NonNull Exception e) {
+                if (e instanceof ResolvableApiException) {
+                    try {
+                        // Show the dialog by calling startResolutionForResult(),
+                        // and check the result in onActivityResult().
+                        ResolvableApiException resolvable = (ResolvableApiException) e;
+                        resolvable.startResolutionForResult(activity,
+                                REQUEST_CHECK_SETTINGS);
+                    } catch (IntentSender.SendIntentException sendEx) {
```

Comment:
What's the consequences for the user when the thing fails? When does it fail?

Reply:
When the intent is canceled or unable to be sent. But this piece of code was copied from [here](https://developer.android.com/training/location/change-location-settings), so I didn't pay much attention to that actually. But this is definitely some non-fatal exception in this case, the user would just have to click the button again.